jetty-http3/http3-qpack/src/main/java/org/eclipse/jetty/http3/qpack/QpackDecoder.java

```diff
@@ -0,0 +1,276 @@
+            LOG.debug("InsertNameReference: nameIndex={}, dynamic={}, value={}", nameIndex, isDynamicTableIndex, value);
+
+        StaticTable staticTable = QpackContext.getStaticTable();
+        DynamicTable dynamicTable = _context.getDynamicTable();
+        Entry referencedEntry = isDynamicTableIndex ? dynamicTable.get(nameIndex) : staticTable.get(nameIndex);
+
+        // Add the new Entry to the DynamicTable.
+        Entry entry = new Entry(new HttpField(referencedEntry.getHttpField().getHeader(), referencedEntry.getHttpField().getName(), value));
+        dynamicTable.add(entry);
+        _instructions.add(new InsertCountIncrementInstruction(1));
+        checkEncodedFieldSections();
+    }
+
+    void insert(String name, String value) throws QpackException
+    {
+        if (LOG.isDebugEnabled())
+            LOG.debug("InsertLiteralEntry: name={}, value={}", name, value);
+
+        Entry entry = new Entry(new HttpField(name, value));
+
+        // Add the new Entry to the DynamicTable.
+        DynamicTable dynamicTable = _context.getDynamicTable();
+        dynamicTable.add(entry);
+        _instructions.add(new InsertCountIncrementInstruction(1));
+        checkEncodedFieldSections();
+    }
+
+    private static int decodeInsertCount(int encInsertCount, int totalNumInserts, int maxTableCapacity) throws QpackException
+    {
+        if (encInsertCount == 0)
+            return 0;
+
+        int maxEntries = maxTableCapacity / 32;
+        int fullRange = 2 * maxEntries;
+        if (encInsertCount > fullRange)
+            throw new QpackException.CompressionException("encInsertCount > fullRange");
+
+        // MaxWrapped is the largest possible value of ReqInsertCount that is 0 mod 2 * MaxEntries.
+        int maxValue = totalNumInserts + maxEntries;
+        int maxWrapped = (maxValue / fullRange) * fullRange;
+        int reqInsertCount = maxWrapped + encInsertCount - 1;
+
+        // If reqInsertCount exceeds maxValue, the Encoder's value must have wrapped one fewer time.
+        if (reqInsertCount > maxValue)
+        {
+            if (reqInsertCount <= fullRange)
+                throw new QpackException.CompressionException("reqInsertCount <= fullRange");
+            reqInsertCount -= fullRange;
+        }
+
+        // Value of 0 must be encoded as 0.
+        if (reqInsertCount == 0)
+            throw new QpackException.CompressionException("reqInsertCount == 0");
+
+        return reqInsertCount;
+    }
+
+    @Override
+    public void dump(Appendable out, String indent) throws IOException
+    {
+        Dumpable.dumpObjects(out, indent, _context.getDynamicTable());
+    }
+
+    @Override
+    public String toString()
+    {
+        return String.format("QpackDecoder@%x{%s}", hashCode(), _context);
+    }
+
+    /**
+     * This delivers notifications from the DecoderInstruction parser directly into the Decoder.
+     */
+    class DecoderAdapter implements DecoderInstructionParser.Handler
```

Comment:
@lachlan-roberts  
I don't think `Adaptor` is the right name here.   That implies it is an abstract implementation that is intended to be further extended.
Rather this is the implementation of the parser handler that integrates into the `Decoder`, so perhaps just call it `InstructionHandler`